### PR TITLE
SSD Handling

### PR DESCRIPTION
**Pull request by Kyep**
After 10-20 minutes of someone being SSD, an extra slot is added for whatever job they have.

This process ignores anyone who is any of the following:
- Not SSD
- Dead
- Off Station
- Anchored, stunned, cuffed, buckled, or pulled
- In space
- A head of staff (including the Captain)
- Not human (ie: a simple mob, borg, etc)
- In a prison area (brig, perma)
- Has gone SSD before in the same round

If you'd like to go SSD for ages without this affecting you, simply buckle yourself to a chair anywhere on the station.

This PR helps avoid situations where SSD people in remote areas of the station are consuming job slots for most of the round, without ever being put in cryo.

As a side bonus, this PR also refactors cryopod.dm to improve it.

🆑 Kyep
rscadd: People who are SSD for awhile will eventually cause more slots to be opened for their job. So someone can do the work they aren't doing. Buckled people, heads, etc are immune.
/🆑

### CODE REVIEW

**Comment by scrubmcnoob** [comment]
Ye, this one is much better than robot cause now it will just auto teleport them to cryo.
And it's simple to avoid this, just buckle yourself like you said.
And hopefully will cut down on people looting SSD players.

**Comment by Aurorablade** [comment]
Still.same thing as not automates.serving that should be human done and/or an aim verb. I am not sold on it.

**Comment by Aurorablade** [comment]
As not rather. Damn phone.

**Comment by scrubmcnoob** [comment]
An admin verb and/or human done? Why not just make this? It's a ton more convenient and will fix a good amount of SSD headache.

**Comment by Aurorablade** [comment]
As bot and mobile is not letting me edit.my comments

**Comment by tigercat2000** [comment]
I strongly disagree with this, for a whole bunch of reasons
 - 5 minutes is not even close to long enough to recover from a total system crash, especially on older machines, it should be 15 minutes or more before they are even considered for being picked up (also, due to the way you made this, people can be picked up literally 2 seconds after becoming SSD, because the controller fires on a static interval)
 - It will absolutely eat aghosted admins, which is a massive fucking issue
 - I don't see this as actually an issue to fix; SSD looting is an administrator issue, and it should be handled by the administrators. We do not want the type of people who flagrantly disregard our rules on the server. I would not be opposed to an alternative for antags, where targets who go SSD for more than, say, 30 minutes, are reassigned. 
 - It makes a big admin notification every time it fires, yay for pointless chat spam every 5 minutes
 - It is blatantly immersion breaking with people just randomly disappearing off the floor in front of other people.

**Comment by Fox-McCloud** [comment]
The concept is better, but I'm still fiercely opposed to this, especially with how debilitating getting cryoghosted is now.

Lose power for a bit->come back and you now have to wait for 2 hours before you can play again; that's really unfair and I've seen even admins be subjected to this (or minor internet blips or small things beyond a player's control).

It seems to me it would be better to just have an automated freeing up of SSD job slots, since that's the real problem.

**Comment by Spacemanspark** [comment]
You mean the system that would automatically kick people if they didn't move for ten minutes?
Fuuuuuuck that system.

**Comment by Kyep** [comment]
@tigercat2000 
- I've changed the timer to 10 minutes. Note that you have to be SSD for two consecutive timers to be picked up. So, a ten minute timer means 5 mins (avg to first timer) plus ten minutes (to second timer) = 15 minutes average time. The way this is written, you CANNOT be eligible for pickup until at least one full cycle has passed. It is NOT possible to go SSD and then get picked up immediately. Even if it was, cryo has its own timeout that adds many more minutes to this.
- It shouldn't eat aghosted admins either, as aghosted admins don't get the 'player_logged' variable set on them while they're aghosted. So, no, that shouldn't be a problem.
- Admins can handle people looting SSDs when they do it. This is about clearing slots, reducing clutter, reducing opportunity, etc. 
- The notification was a debug thing I accidentally left in, its deleted now.
- The immersion breaking thing was my primary reason why I made an SSD bot. But that got shot down in flames. Fox in particular gave the impression that the main reason he did not like it was that it was a bot and not a controller. So, now its a controller.

@Fox-McCloud 
- Automated freeing of job slots is possible, it would use almost all the same code. It wouldn't help with SSDs though. I still prefer to address the problem at source and cryo the SSDs.
- @Spacemanspark This is not an AFK kicker, nor is it intended to be. This is to cryo people who've left the round and aren't coming back. It has a substantial timer before it even picks them up, and then there is the cryo timer on top of that. If people are so happy to have admins handle SSDs manually, why don't we have the system auto-cryo SSDs, and then admins can handle the really rare cases where someone's power fails for 30 minutes, by respawning them? That's way more efficient (automate the handling of the simple, common, non-problematic case, and have manual intervention handle the tough ones) than the alternative (have even the basic cases require active intervention).

**Comment by tigercat2000** [comment]
Oh, I see how you wrote this to prevent the immediate pickup. Kinda messy.

Still don't like this, though, I view it as wholey unnecessary, and there are better alternatives.

**Comment by Kyep** [comment]
Freeing their job slot instead of cryoing them is one way to go. 99% of the code would be reusable for that function. But it still feels like applying hair dye rather than tackling the root of the problem.

**Comment by Fox-McCloud** [comment]
It's not really addressing the problem at the source though; the problem is the player going SSD to begin with. This isn't a problem we can inherently resolve, at its core, unfortunately---we can just mitigate the negative effects of it happening.

The most negative effects are hogging up job slots, thereby limiting the ability of newer players to enter the round as a job besides security or civilian. To that end, your PR does resolve that, but it also introduces a lot of other problems and can potentially hurt players who disconnected and hare having trouble reconnecting as opposed to intending to go permanently SSD.

**Comment by scrubmcnoob** [comment]
This is implying that our players who go SSD from a system crash are unable to get onto the server, at the very least, within ten minutes.
I understand booting an old system up may take longer but to just launch byond and open the game? I can't imagine this game bogs down PCs too much. And this is assuming most of our players are using toasters.

**Comment by Fox-McCloud** [comment]
It's not just crashes---it's internet and electrical blips; even some of our admins have routine problems with this, where they'll be gone for 15-30 minutes, but have every intention of returning.

Those individuals shouldn't be punished for having something happen to them beyond their control--now they may have to wait upwards of 1.5 hours to play again because of this system.

**Comment by scrubmcnoob** [comment]
If this is an issue that affects many players, and would be more con than pro for our server, then we should just do the job slot cleanup thing and simply let admins manually deal with people touching SSD.
But if it isn't, then I am all for just automatically cleaning SSDs.
Which ever option is more beneficial to the server, really.

**Comment by AndriiYukhymchak** [comment]
1st people can still move them to the cryopods by hand, so all arguments about "being unfair to system crashes" is a bit odd for me.
But really it has to move them from target list and from job list.
There is still way to abuse it, like cryoing self to remove self from all traitors objective, but that can be mitigated...
For example if person that was a target disconnects, waits 15 minutes and then reconnects, every traitor should get optional objective to eliminate them (if such objective existed)

**Comment by Aurorablade** [comment]
it dosen't need to be automated. It needs to be something done by target admin verb on a player. Until that point is reached i cannot support this PR.

**Comment by Fox-McCloud** [comment]
>People can still move them to the cryopods by hand, so all arguments about "being unfair to system crashes" is a bit odd for me.

This has agency though; often times it's very obvious to the players that someone has abruptly disconnected. To the controller, it won't care; it'll just poof them away no matter what.

When I'm having a conversation with someone or someone is doing their job *or* it's a player who normally sticks around and they suddenly poof, out of the blue...my gut reaction isn't to go "welp, time to cryo them"; it's "welp, going to buckle them to a chair, they'll probably be back soon enough". 

Again, a controller can't make these kinds of judgment calls; players can (and do).

**Comment by scrubmcnoob** [comment]
Wouldn't those buckled people not get poofed away because the controller checks for that?
Again, I'm all for any system that helps with job slots.

**Comment by Twinmold93** [comment]
Yet if someone cryos them anyway, they'll still have to wait anyway. The only difference is this automatics it. Most people wait 10 to 15 minutes before sending to cryo, then that person sits in cryo another 15 minutes. If they aren't back by then, they get sent to storage. People can still mitigate it by just buckling the person to a chair if they know they are going to come back. If that person is off on their own, not interacting with people, etc, they get sent away.

The big issue is actually any ITEMS they had on them. There are items people have that are in limited supply for a good number of roles, yet moving them straight to cryo means the loss of those items (or requiring someone to actively watch for cryo items anyway). What would be the solution to this problem?

**Comment by Draconic-saint** [comment]
Eliminating the human element in this is a no no, strong thumbs down

**Comment by Anticept** [comment]
For SSD looting: why not just DISALLOW a person to open inventory on someone who is SSD in the code itself? that seems like such an incredibly simple way to put an end to that and drop admin workload considerably.

And, if the cryoconsole would be FIXED so that it would log ALL items instead of just a few fluff ones (it doesn't even save the captains gloves!), that solves a huge number of issues revolving around SSDs in the first place!

As for job slots: there isn't a clean way of doing this without cryoing the player. What happens if they are gone for half an hour and come back? Do they get their old job? I guess its not really a problem unless you are dealing with a head role.

**Comment by tigercat2000** [comment]
@Anticept 
>Get in fight
>drop connection
>lol can't do anything but drag you to cryo
>> then they space your corpse

**Comment by Anticept** [comment]
That's already something admins step in on with an ahelp.

**Comment by scrubmcnoob** [comment]
> That's already something admins step in on with an ahelp. Guy goes SSD, you can't attack them anyways.

Actually no @Anticept If someone is being murdered by an antag and want's to live, they just go SSD and can claim 'lol game crash'
The concept of forcing immunity on SSDs via code will cause more problems than it will fix.

**Comment by Anticept** [comment]
Then the rule needs to change. I've had actual game crashes during a fight and while I felt salty, I took it with grace. Crashes during a fight, for me anyways, have only happened a couple times out of all the fights I've been in.

**Comment by scrubmcnoob** [comment]
> Then the rule needs to change.
@Anticept To what?
You can't murder a person if they go randomly SSD while your cuffing them as a vamp?
Or you can't go SSD during a fight?
What happens if the person actually crashed?
Again, causes plenty of more problems than fixes.

**Comment by Anticept** [comment]
Aren't these all issues that could easily be resolved with an ahelp though anyways? As it stands, everytime an interaction is happening and the person goes SSD suddenly, it has to be ahelped. This is also on top of people already looting and attacking SSDs anyways, which is what the idea serves to stop.

It had been such a problem that ansari has had to use server announcements and hand out bans to curb the greytiding behavior on SSDs

That said: alternatively, generate an admin log when someone attacks or *finishes* looting an SSD?

**Comment by Aurorablade** [comment]
it already is logged.

**Comment by Birdtalon** [comment]
Looting and attacking SSDs like @tigercat2000 said earlier is an admin issue and shouldn't be handled by any kind of automatic process. If someone blatantly disregards the rules and continues to abuse SSD players then I'd rather it be picked up an manually dealt with by administrators so the correct sanctions can be imposed.

Also, through just playing the game I often find the majority of the players who are SSD are infact; civilians, therefore very few job slots would actually be freed up by this process. Certainly no more than through our experienced players taking the SSD ones to cryo. As @Fox-McCloud said earlier, players are perfectly capable of making the decision to cryo people and when to do it. Our current server is limited more by lack of jobslots and content than by SSDs taking the slots. Which is a whole different issue entirely.

**Comment by Anticept** [comment]
I guess I'm missing the point of letting it continue to be handled manually when admins are already significantly overburdened with shitty behavior. Coming up with ways to mitigate the problem just feels more productive than just giving up on the whole idea because of a few bumps to consider and get past.

Or hopefully more admins are onboarded in the near future to help cope with the workload.

**Comment by Kyep** [comment]
@Fox-McCloud @Crazylemon64 
PR changed. Now, it only frees the job slots. It does not cryo people.

**Comment by Necaladun** [comment]
Will this cause a doubling up? Eg, someone who is SSD has their job slot freed, then is put in cryo, freeing the job slot again?

**Comment by Aurorablade** [comment]
And that's still not fair to people who intend to come back. I maintain that this is better handled on a human level then code.

**Comment by Kyep** [comment]
@Necaladun Nope. The mind.role_given_up exists to prevent that from happening.
@Aurorablade Meh. IMHO, if you disconnect from the game for enough minutes, I think it should be expected that you give up your job slot. Slots are a limited resource, and there should be a limit to how long someone who is not even playing can keep one reserved for them.

File: code/game/machinery/cryopod.dm [hunk 1]
```diff
@@ -481,6 +480,31 @@
 			//Despawning occurs when process() is called with an occupant without a client.
 			src.add_fingerprint(M)
 
+/obj/machinery/cryopod/proc/take_occupant(var/mob/living/carbon/E, var/willing_factor = 1)
```
Comment (Fox-McCloud) [inline code/game/machinery/cryopod.dm]: You amde this proc, but yet left copy-pasta of it in.

File: code/game/machinery/cryopod.dm [hunk 2]
```diff
@@ -481,6 +480,31 @@
...
+	if(src.occupant)
```
Comment (Fox-McCloud) [inline code/game/machinery/cryopod.dm]: implied `src`

File: code/game/machinery/cryopod.dm [hunk 3]
```diff
@@ -481,6 +480,31 @@
...
+	time_entered = world.time
+	if(findtext("[E.key]","@",1,2))
+		var/FT = replacetext(E.key, "@", "")
+		for(var/mob/dead/observer/Gh in respawnable_list) //this may not be foolproof but it seemed like a better option than 'in world'
+			if(Gh.key == FT)
+				if(Gh.client && Gh.client.holder) //just in case someone has a byond name with @ at the start, which I don't think is even possible but whatever
+					to_chat(Gh, "<span style='color: #800080;font-weight: bold;font-size:4;'>Warning: Your body has entered cryostorage.</span>")
+	log_admin("<span class='notice'>[key_name(E)] entered a stasis pod.</span>")
+	message_admins("[key_name_admin(E)] entered a stasis pod. (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[x];Y=[y];Z=[z]'>JMP</a>)")
+	src.add_fingerprint(E)
```
Comment (Fox-McCloud) [inline code/game/machinery/cryopod.dm]: implied `src`

File: code/controllers/Processes/ssd.dm [hunk 4]
```diff
@@ -0,0 +1,57 @@
+
+var/global/datum/controller/process/ssd/ssd
```
Comment (Fox-McCloud) [inline code/controllers/Processes/ssd.dm]: standard naming connection dicates this should be `ssd_master`

File: code/controllers/Processes/ssd.dm [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		return 0
+	if(!isLivingSSD(src))
+		return 0
+	if(anchored)
+		return 0
+	if(stunned)
+		return 0
+	if(handcuffed)
+		return 0
+	if(istype(get_turf(src), /turf/space))
```
Comment (Fox-McCloud) [inline code/controllers/Processes/ssd.dm]: there's an istype helper for this

File: code/controllers/Processes/ssd.dm [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	var/list/ignored_ssds = list()
+	var/list/areas_to_ignore = list(/area/security/prison)
+
+/datum/controller/process/ssd/setup()
+	name = "SSD"
+	schedule_interval = 6000
+
+/datum/controller/process/ssd/doWork()
+	current_ssds = list()
+	for(var/mob/living/carbon/human/H in mob_list)
```
Comment (Fox-McCloud) [inline code/controllers/Processes/ssd.dm]: why loop through the entire mob list if the dead SSDs return 0 and aren't elligible for job slot clearing?

**Comment by Fox-McCloud** [comment]
The controller comes off as a bit overcomplicated, at first glance--it's also unreliable.

If someone goes SSD right before it fires, then their job slot will be freed in 10 minutes.

If they go SSD right *after* it fires, then it'll be 20 minutes before their job slot is freed.

This isn't good; it should be something that can be consistently relied upon. It would probably be better if human mobs had a var that marked when they went SSD. If the controller sees that they've been SSD for over 20 minutes, then it will remove the job.

You'd have to then lower how often the controller fired (probably every 1-2 minutes then), but then you don't have to do so much looping and complication in the controller itself. Yeah, there's still variation (20-22 minutes), but you don't have a massive 10-20 minute "may get get your job freed or not" during that time period.

**Pull request closed without merging**